python/ray/util/ray_executor.py

```diff
@@ -0,0 +1,237 @@
+                results = self._map(self.submit, fn, chunk, timeout)
+            results_list.append(results)
+        return itertools.chain(*results_list)
+
+    @staticmethod
+    def _get_chunks(*iterables, chunksize):
+        """
+        https://github.com/python/cpython/blob/main/Lib/concurrent/futures/process.py#L186
+        Iterates over zip()-ed iterables in chunks.
+        """
+        it = zip(*iterables)
+        while True:
+            chunk = tuple(itertools.islice(it, chunksize))
+            if not chunk:
+                return
+            yield chunk
+
+    @staticmethod
+    def _result_or_cancel(fut, timeout=None):
+        """
+        From concurrent.futures
+        """
+        try:
+            try:
+                return fut.result(timeout)
+            finally:
+                fut.cancel()
+        finally:
+            # Break a reference cycle with the exception in self._exception
+            del fut
+
+    def _map(self, submit_fn, fn, iterables, timeout=None):
+        """
+        This was adapted from concurrent.futures.Executor.map.
+        """
+        if timeout is not None:
+            end_time = timeout + time.monotonic()
+
+        fs = [submit_fn(fn, *args) for args in iterables]
+
+        # Yield must be hidden in closure so that the futures are submitted
+        # before the first iterator value is required.
+        def result_iterator():
+            try:
+                # reverse to keep finishing order
+                fs.reverse()
+                while fs:
+                    # Careful not to keep a reference to the popped future
+                    if timeout is None:
+                        yield self._result_or_cancel(fs.pop())
+                    else:
+                        yield self._result_or_cancel(
+                            fs.pop(), end_time - time.monotonic()
+                        )
+            finally:
+                for future in fs:
+                    future.cancel()
+
+        return result_iterator()
+
+    def shutdown(self, wait=True, *, cancel_futures=False):
+        """Clean-up the resources associated with the Executor.
+
+        It is safe to call this method several times. No other methods can be
+        called after this one.
+
+        Args:
+            wait: If True then shutdown will not return until all running
+                futures have finished executing and the resources used by the
+                executor have been reclaimed.
+            cancel_futures: If True then shutdown will cancel all pending
+                futures. Futures that are completed or running will not be
+                cancelled.
+        """
+        self._shutdown_lock = True
+
+        if cancel_futures:
+            for future in self._futures:
+                _ = future.cancel()
+
+        if wait:
+            for future in self._futures:
+                if future.running():
+                    _ = future.result()
+
+        ray.shutdown()
```

Comment:
I think this part can be contraversial. This also means if you use this API with `with` statement, you will create/destroy ray clusters again and again (or keep recreating new jobs since you do ray.init() -> ray.shutdown repeatedly). Why don't we just have a separate argument `shutdown_ray=False` and we shutdown when it is only True? 

If we go with this approach, we should also destroy actors and futures upon shutdown. So we should delete all `self._futures` and `self. __actor_pool `

Reply:
We have opted to add a `shutdown_ray=True` argument. This way the executor can still be used as a drop-in replacement for other Executors, but users have the option to set the arg to False and preserve the Ray instance if they choose to.

Reply:
I have a proposal we can both satisfy/

1. If ray is already initialized, we don't shutdown.
2. If not, we shutdown (in this case, it is a drop in replacement)